staging/src/k8s.io/kube-aggregator/pkg/controllers/status/available_controller.go

```diff
@@ -449,33 +455,49 @@ func (c *AvailableConditionController) deleteAPIService(obj interface{}) {
 		}
 	}
 	klog.V(4).Infof("Deleting %q", castObj.Name)
-	c.enqueue(castObj)
+	if castObj.Spec.Service != nil {
+		c.rebuildAPIServiceCache()
+	}
+	c.enqueue(castObj.Name)
 }
 
-// there aren't very many apiservices, just check them all.
-func (c *AvailableConditionController) getAPIServicesFor(obj runtime.Object) []*apiregistration.APIService {
+func (c *AvailableConditionController) getAPIServicesFor(obj runtime.Object) []string {
 	metadata, err := meta.Accessor(obj)
 	if err != nil {
 		utilruntime.HandleError(err)
 		return nil
 	}
+	c.RLock()
+	defer c.RUnlock()
+	if svcMap, ok := c.cache[metadata.GetNamespace()]; ok {
+		if names, exist := svcMap[metadata.GetName()]; exist {
+			return names
+		}
+	}
+	return nil
+}
 
-	var ret []*apiregistration.APIService
+func (c *AvailableConditionController) rebuildAPIServiceCache() {
 	apiServiceList, _ := c.apiServiceLister.List(labels.Everything())
+	c.Lock()
+	defer c.Unlock()
 	for _, apiService := range apiServiceList {
 		if apiService.Spec.Service == nil {
 			continue
```

Comment:
if this apiservice previously had a non-nil service, it would need to be removed from the cache. rather than trying to mutate the cache in-place, it would be much simpler to rebuild the cache from scratch, then overwrite `c.cache`